source/common/router/rds_impl.cc

```diff
@@ -51,110 +52,58 @@ RouteConfigProviderSharedPtr RouteConfigProviderUtil::create(
 }
 
 StaticRouteConfigProviderImpl::StaticRouteConfigProviderImpl(
-    const envoy::config::route::v3::RouteConfiguration& config,
-    const OptionalHttpFilters& optional_http_filters,
+    const envoy::config::route::v3::RouteConfiguration& config, Rds::ConfigTraits& config_traits,
     Server::Configuration::ServerFactoryContext& factory_context,
-    ProtobufMessage::ValidationVisitor& validator,
-    RouteConfigProviderManagerImpl& route_config_provider_manager)
-    : config_(new ConfigImpl(config, optional_http_filters, factory_context, validator, true)),
-      route_config_proto_{config}, last_updated_(factory_context.timeSource().systemTime()),
-      route_config_provider_manager_(route_config_provider_manager) {
-  route_config_provider_manager_.static_route_config_providers_.insert(this);
-}
+    Rds::RouteConfigProviderManager& route_config_provider_manager)
+    : base_(config, config_traits, factory_context, route_config_provider_manager),
+      route_config_provider_manager_(route_config_provider_manager) {}
 
 StaticRouteConfigProviderImpl::~StaticRouteConfigProviderImpl() {
-  route_config_provider_manager_.static_route_config_providers_.erase(this);
+  route_config_provider_manager_.eraseStaticProvider(this);
+}
+
+ConfigConstSharedPtr StaticRouteConfigProviderImpl::configCast() {
+  ASSERT(dynamic_cast<const Config*>(base_.config().get()));
+  return std::static_pointer_cast<const Config>(base_.config());
```

Comment:
Why can't we use inheritance for the `Config` objects here? It doesn't work for proto, since these are generated types, but I would think we could have a common C++ `Config` hierarchy.

Reply:
We are using inheritance. The `Envoy::Router::Config` is inherited from `Envoy::Rds::Config` what is passed back by `base_.config()`. The problem is that is an empty class to just be able to do the downcast.

The `Config` classes have basically only a `route()` function but the parameters and the return value are quite protocol specific, for example:
Thift:
`virtual RouteConstSharedPtr route(const MessageMetadata& metadata, uint64_t random_value) const PURE;`
Http:
`virtual RouteConstSharedPtr route(const Http::RequestHeaderMap& headers, const StreamInfo::StreamInfo& stream_info, uint64_t random_value) const PURE; `

Reply:
OK, makes sense, thanks for clarifying.